server.js

```diff
@@ -2056,14 +2047,57 @@ if (config.startServer) {
           process.exit(0);
         }
 
-        process.on('SIGTERM', () => {
-          opentelemetry
-            .shutdown()
-            .catch((err) => logger.error('Error shutting down OpenTelemetry', err))
-            .finally(() => {
-              process.exit(0);
-            });
-        });
+        const prepareForTermination = async () => {
+          // We want to proceed with termination even if something goes wrong,
+          // so don't allow this function to throw.
+          try {
+            await util.promisify(server.close).call(server);
+
+            await socketServer.close();
+
+            // Wait for all currently-executing cron jobs to finish.
+            await cron.stop();
+
+            // Wait for all currently-executing server jobs to finish.
+            await serverJobs.stop();
+          } catch (err) {
+            logger.error('Error while shutting down server', err);
+            Sentry.captureException(err);
+          }
+        };
+
+        const terminate = async () => {
+          // Shut down OpenTelemetry exporting.
+          try {
+            opentelemetry.shutdown();
+          } catch (err) {
+            logger.error('Error shutting down OpenTelemetry', err);
+            Sentry.captureException(err);
+          }
+
+          // Flush all events to Sentry.
+          try {
+            await Sentry.flush();
+          } finally {
+            process.exit(0);
+          }
+        };
+
+        // If we're running in EC2 auto scaling, this will wait for our instance
+        // to enter the `Terminating:Wait` state.
+        //
+        // If we're not running in EC2, this will be a no-op.
+        lifecycleHooks
+          .handleAndCompleteInstanceTermination(prepareForTermination)
+          .catch((err) => {
+            logger.error('Error handling instance termination', err);
+            Sentry.captureException(err);
+          })
+          .finally(terminate);
```

Comment:
Not that I particularly object to the promise syntax, but I'm wondering whether we should try and use async/await syntax in cases like this? Mainly just to standardize on a/a to avoid having multiple equivalent syntaxes. Of course I don't mind using promises explicitly where it's much simpler to do so.

Reply:
I typically use `.then()`/`.catch()`/`.finally()` only when I don't want to actually wait for something to complete before proceeding to the next statement. In this case, we need to keep executing so that we could attach the `SIGTERM` handler. The alternative here would be an IIFE, which IMO is worse:

```js
(async () => {
  try {
    await lifecycleHooks.handleAndCompleteInstanceTermination(prepareForTermination);
  } catch (err) {
    logger.error('Error handling instance termination', err);
    Sentry.captureException(err);
  }

  await terminate();
})();
```

Reply:
Yeah, I agree that an IIFE is pretty nasty and promises are better here.

Reply:
(and thanks for the detailed explanation)